src/Forum/AuthenticationResponseFactory.php

```diff
@@ -12,107 +12,79 @@
         }
 
-        $user = User::where($identification)->first();
-
-        $payload = $this->getPayload($identification, $suggestions, $user);
+        if ($user = User::where($data['identification'])->first()) {
+            $payload = ['authenticated' => true];
+        } else {
+            $token = $this->generateToken($data['identification'], $data['attributes'] ?? []);
+            $payload = $this->buildPayload($token, $data['identification'], $data['suggestions'] ?? []);
+        }
 
-        $response = $this->getResponse($payload);
+        $response = $this->makeResponse($payload);
 
         if ($user) {
-            $session = $request->getAttribute('session');
-            $this->authenticator->logIn($session, $user->id);
-
             $response = $this->rememberer->rememberUser($response, $user->id);
         }
 
         return $response;
     }
 
-    /**
-     * @param string $username
-     * @return string
-     */
-    private function sanitizeUsername($username)
+    private function sanitizeUsername(string $username): string
     {
         return preg_replace('/[^a-z0-9-_]/i', '', $username);
     }
 
-    /**
-     * @param array $payload
-     * @return HtmlResponse
-     */
-    private function getResponse(array $payload)
+    private function generateToken(array $identification, array $attributes): AuthToken
     {
-        $content = sprintf(
-            '<script>window.opener.app.authenticationComplete(%s); window.close();</script>',
-            json_encode($payload)
-        );
+        $payload = array_merge($identification, $attributes);
 
-        return new HtmlResponse($content);
+        $token = AuthToken::generate($payload);
+        $token->save();
+
+        return $token;
     }
 
-    /**
-     * @param array $identification
-     * @param array $suggestions
-     * @param User|null $user
-     * @return array
-     */
-    private function getPayload(array $identification, array $suggestions, User $user = null)
+    private function buildPayload(AuthToken $token, array $identification, array $suggestions): array
     {
-        // If a user with these attributes already exists, then we will log them
-        // in by generating an access token. Otherwise, we will generate a
-        // unique token for these attributes and add it to the response, along
-        // with the suggested account information.
-        if ($user) {
-            $payload = ['authenticated' => true];
-        } else {
-            $token = AuthToken::generate($identification);
-            $token->save();
-
-            $payload = array_merge(
-                $identification,
-                $suggestions,
-                ['token' => $token->id],
-                // List of the fields that can't be edited during sign up
-                ['identificationFields' => array_keys($identification)]
-            );
-        }
+        $payload = array_merge($identification, $suggestions);
```

Comment:
Not sure if this is critical at all, but I could imagine that preferring the keys in `$identification` over those in `$suggestions` (by swapping the parameter order here) might be somewhat safer?